tabu.sty

```diff
@@ -1298,6 +1299,58 @@
 \def\tabu@endofcollect #1{\ifnum0=`{}\fi
                           \expandafter\endgroup \the\toks@  #1%
 }% \tabu@endofcollect
+\catcode`\| \z@
+\catcode`\\ 12
+|long|def|tabu@collectbody@forscan #1#2#3\end{%
+    |def|tabu@collectbody@forscan@next{|tabu@collectbody@forscan{#1}}%
+    |def|tabu@collectbody@forscan@partial{#2#3}%
+    |futurelet|tabu@temp@token|tabu@collectbody@forscan@a
+}%
+|def|tabu@collectbody@forscan@a{%
+    |ifcat|tabu@temp@token e%
+        |expandafter|@firstoftwo
+    |else
+        |expandafter|@secondoftwo
+    |fi
+    |tabu@collectbody@forscan@aa
+    |tabu@collectbody@forscan@b
+}%
+|def|tabu@collectbody@forscan@aa #1{%
+    |expandafter|tabu@collectbody@forscan@next|expandafter{|tabu@collectbody@forscan@partial\end#1}%
```

Comment:
I document it now before I forget: here we had some `\endfoo`, thus it was started by some `\foo`. Hence there is no associated  `\begin`. Good thing LaTeX does not have `\beginfoo` syntax... Thus the balancing done via `\tabu@pushbegins@forscan` isn't offset, and the loop terminates correctly after tabu body is collected. Definitely this won't work with `\endtabu`, but original surely did not either as it is checking for `\end` token.